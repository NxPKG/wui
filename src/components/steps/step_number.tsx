/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { FunctionComponent, HTMLAttributes } from 'react';
import classNames from 'classnames';

import { WuiIcon } from '../icon';

import { WuiStepProps } from './step';

import { WuiI18n } from '../i18n';
import { CommonProps, keysOf } from '../common';

const statusToClassNameMap = {
  complete: 'wuiStepNumber--complete',
  incomplete: 'wuiStepNumber--incomplete',
  warning: 'wuiStepNumber--warning',
  danger: 'wuiStepNumber--danger',
  disabled: 'wuiStepNumber--disabled',
};

export const STATUS = keysOf(statusToClassNameMap);

export type WuiStepStatus =
  | 'complete'
  | 'incomplete'
  | 'warning'
  | 'danger'
  | 'disabled';

export interface WuiStepNumberProps
  extends CommonProps,
    HTMLAttributes<HTMLDivElement> {
  /**
   * May replace the number provided in props.number with alternate styling
   */
  status?: WuiStepStatus;
  number?: number;
  /**
   * Uses a border and removes the step number
   */
  isHollow?: boolean;
  /**
   * Title sizing equivalent to WuiTitle, but only `m`, `s` and `xs`. Defaults to `s`
   */
  titleSize?: WuiStepProps['titleSize'];
}

export const WuiStepNumber: FunctionComponent<WuiStepNumberProps> = ({
  className,
  status,
  number,
  isHollow,
  titleSize,
  ...rest
}) => {
  const classes = classNames(
    'wuiStepNumber',
    status ? statusToClassNameMap[status] : undefined,
    {
      'wuiStepNumber-isHollow': isHollow,
    },
    className
  );

  const iconSize = titleSize === 'xs' ? 's' : 'm';

  let numberOrIcon;
  if (status === 'complete') {
    numberOrIcon = (
      <WuiI18n token="wuiStepNumber.isComplete" default="complete">
        {(isComplete: string) => (
          <WuiIcon
            type="check"
            className="wuiStepNumber__icon"
            size={iconSize}
            aria-label={isComplete}
          />
        )}
      </WuiI18n>
    );
  } else if (status === 'warning') {
    numberOrIcon = (
      <WuiI18n token="wuiStepNumber.hasWarnings" default="has warnings">
        {(hasWarnings: string) => (
          <WuiIcon
            type="alert"
            className="wuiStepNumber__icon"
            size={iconSize}
            aria-label={hasWarnings}
          />
        )}
      </WuiI18n>
    );
  } else if (status === 'danger') {
    numberOrIcon = (
      <WuiI18n token="wuiStepNumber.hasErrors" default="has errors">
        {(hasErrors: string) => (
          <WuiIcon
            type="cross"
            className="wuiStepNumber__icon"
            size={iconSize}
            aria-label={hasErrors}
          />
        )}
      </WuiI18n>
    );
  } else if (!isHollow) {
    numberOrIcon = number;
  }

  return (
    <div className={classes} {...rest}>
      {numberOrIcon}
    </div>
  );
};
