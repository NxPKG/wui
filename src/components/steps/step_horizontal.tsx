/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  FunctionComponent,
  HTMLAttributes,
  MouseEventHandler,
} from 'react';
import { CommonProps } from '../common';
import classNames from 'classnames';

import { WuiI18n } from '../i18n';
import { WuiScreenReaderOnly, WuiKeyboardAccessible } from '../accessibility';

import { WuiStepStatus, WuiStepNumber } from './step_number';

export interface WuiStepHorizontalProps
  extends CommonProps,
    HTMLAttributes<HTMLDivElement> {
  /**
   * Is the current step
   */
  isSelected?: boolean;
  /**
   * Is a previous step that has been completed
   */
  isComplete?: boolean;
  onClick: MouseEventHandler<HTMLDivElement>;
  disabled?: boolean;
  /**
   * The number of the step in the list of steps
   */
  step?: number;
  title?: string;
  /**
   * May replace the number provided in props.step with alternate styling.
   * The `isSelected`, `isComplete`, and `disabled` props will override these.
   */
  status?: WuiStepStatus;
}

export const WuiStepHorizontal: FunctionComponent<WuiStepHorizontalProps> = ({
  className,
  step = 1,
  title,
  isSelected,
  isComplete,
  onClick,
  disabled,
  status,
  ...rest
}) => {
  const classes = classNames('wuiStepHorizontal', className, {
    'wuiStepHorizontal-isSelected': isSelected,
    'wuiStepHorizontal-isComplete': isComplete,
    'wuiStepHorizontal-isIncomplete': !isSelected && !isComplete,
    'wuiStepHorizontal-isDisabled': disabled,
  });

  if (disabled) {
    status = 'disabled';
  } else if (isComplete) {
    status = 'complete';
  } else if (isSelected) {
    status = status;
  } else if (!status) {
    status = 'incomplete';
  }

  const onStepClick = (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    if (disabled) return;
    onClick(event);
  };

  return (
    <WuiI18n
      token="wuiStepHorizontal.buttonTitle"
      default={({
        step,
        title,
        disabled,
        isComplete,
      }: Pick<
        WuiStepHorizontalProps,
        'step' | 'title' | 'disabled' | 'isComplete'
      >) => {
        let titleAppendix = '';
        if (disabled) {
          titleAppendix = ' is disabled';
        } else if (isComplete) {
          titleAppendix = ' is complete';
        }

        return `Step ${step}: ${title}${titleAppendix}`;
      }}
      values={{ step, title, disabled, isComplete }}>
      {(buttonTitle: string) => (
        <WuiKeyboardAccessible>
          <div
            role="tab"
            aria-selected={!!isSelected}
            aria-disabled={!!disabled}
            className={classes}
            onClick={onStepClick}
            tabIndex={disabled ? -1 : 0}
            title={buttonTitle}
            {...rest}>
            <WuiScreenReaderOnly>
              <div>
                <WuiI18n token="wuiStepHorizontal.step" default="Step" />
              </div>
            </WuiScreenReaderOnly>

            <WuiStepNumber
              className="wuiStepHorizontal__number"
              status={status}
              number={step}
            />

            <div className="wuiStepHorizontal__title">{title}</div>
          </div>
        </WuiKeyboardAccessible>
      )}
    </WuiI18n>
  );
};
