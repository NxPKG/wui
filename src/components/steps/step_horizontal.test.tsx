/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, mount } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { STATUS } from './step_number';
import { WuiStepHorizontal } from './step_horizontal';

describe('WuiStepHorizontal', () => {
  test('is rendered', () => {
    const component = render(
      <WuiStepHorizontal {...requiredProps} onClick={() => {}} />
    );

    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    test('step', () => {
      const component = render(
        <WuiStepHorizontal step={5} onClick={() => {}} />
      );

      expect(component).toMatchSnapshot();
    });

    test('title', () => {
      const component = render(
        <WuiStepHorizontal title={'First step'} onClick={() => {}} />
      );

      expect(component).toMatchSnapshot();
    });

    test('isSelected', () => {
      const component = render(
        <WuiStepHorizontal isSelected onClick={() => {}} />
      );

      expect(component).toMatchSnapshot();
    });

    test('isComplete', () => {
      const component = render(
        <WuiStepHorizontal isComplete onClick={() => {}} />
      );

      expect(component).toMatchSnapshot();
    });

    describe('status', () => {
      STATUS.forEach(status => {
        test(`${status} is rendered`, () => {
          const component = render(
            <WuiStepHorizontal status={status} onClick={() => {}} />
          );

          expect(component).toMatchSnapshot();
        });
      });
    });

    describe('onClick', () => {
      test('is called when clicked', () => {
        const onClickHandler = jest.fn();

        const component = mount(
          <WuiStepHorizontal step={1} onClick={onClickHandler} />
        );

        component.simulate('click');

        expect(onClickHandler).toBeCalledTimes(1);
      });

      test("isn't called when clicked if it's disabled", () => {
        const onClickHandler = jest.fn();

        const component = mount(
          <WuiStepHorizontal disabled step={1} onClick={onClickHandler} />
        );

        component.simulate('click');

        expect(onClickHandler).not.toBeCalled();
      });
    });
  });
});
