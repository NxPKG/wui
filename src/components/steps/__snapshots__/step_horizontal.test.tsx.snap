// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WuiStepHorizontal is rendered 1`] = `
<div
  aria-disabled="false"
  aria-label="aria-label"
  aria-selected="false"
  class="wuiStepHorizontal testClass1 testClass2 wuiStepHorizontal-isIncomplete"
  data-test-subj="test subject string"
  role="tab"
  tabindex="0"
  title="Step 1: undefined"
>
  <div
    class="wuiScreenReaderOnly"
  >
    Step
  </div>
  <div
    class="wuiStepNumber wuiStepNumber--incomplete wuiStepHorizontal__number"
  >
    1
  </div>
  <div
    class="wuiStepHorizontal__title"
  />
</div>
`;

exports[`WuiStepHorizontal props isComplete 1`] = `
<div
  aria-disabled="false"
  aria-selected="false"
  class="wuiStepHorizontal wuiStepHorizontal-isComplete"
  role="tab"
  tabindex="0"
  title="Step 1: undefined is complete"
>
  <div
    class="wuiScreenReaderOnly"
  >
    Step
  </div>
  <div
    class="wuiStepNumber wuiStepNumber--complete wuiStepHorizontal__number"
  >
    <div
      aria-label="complete"
      class="wuiStepNumber__icon"
      data-wuiicon-type="check"
    />
  </div>
  <div
    class="wuiStepHorizontal__title"
  />
</div>
`;

exports[`WuiStepHorizontal props isSelected 1`] = `
<div
  aria-disabled="false"
  aria-selected="true"
  class="wuiStepHorizontal wuiStepHorizontal-isSelected"
  role="tab"
  tabindex="0"
  title="Step 1: undefined"
>
  <div
    class="wuiScreenReaderOnly"
  >
    Step
  </div>
  <div
    class="wuiStepNumber wuiStepHorizontal__number"
  >
    1
  </div>
  <div
    class="wuiStepHorizontal__title"
  />
</div>
`;

exports[`WuiStepHorizontal props status complete is rendered 1`] = `
<div
  aria-disabled="false"
  aria-selected="false"
  class="wuiStepHorizontal wuiStepHorizontal-isIncomplete"
  role="tab"
  tabindex="0"
  title="Step 1: undefined"
>
  <div
    class="wuiScreenReaderOnly"
  >
    Step
  </div>
  <div
    class="wuiStepNumber wuiStepNumber--complete wuiStepHorizontal__number"
  >
    <div
      aria-label="complete"
      class="wuiStepNumber__icon"
      data-wuiicon-type="check"
    />
  </div>
  <div
    class="wuiStepHorizontal__title"
  />
</div>
`;

exports[`WuiStepHorizontal props status danger is rendered 1`] = `
<div
  aria-disabled="false"
  aria-selected="false"
  class="wuiStepHorizontal wuiStepHorizontal-isIncomplete"
  role="tab"
  tabindex="0"
  title="Step 1: undefined"
>
  <div
    class="wuiScreenReaderOnly"
  >
    Step
  </div>
  <div
    class="wuiStepNumber wuiStepNumber--danger wuiStepHorizontal__number"
  >
    <div
      aria-label="has errors"
      class="wuiStepNumber__icon"
      data-wuiicon-type="cross"
    />
  </div>
  <div
    class="wuiStepHorizontal__title"
  />
</div>
`;

exports[`WuiStepHorizontal props status disabled is rendered 1`] = `
<div
  aria-disabled="false"
  aria-selected="false"
  class="wuiStepHorizontal wuiStepHorizontal-isIncomplete"
  role="tab"
  tabindex="0"
  title="Step 1: undefined"
>
  <div
    class="wuiScreenReaderOnly"
  >
    Step
  </div>
  <div
    class="wuiStepNumber wuiStepNumber--disabled wuiStepHorizontal__number"
  >
    1
  </div>
  <div
    class="wuiStepHorizontal__title"
  />
</div>
`;

exports[`WuiStepHorizontal props status incomplete is rendered 1`] = `
<div
  aria-disabled="false"
  aria-selected="false"
  class="wuiStepHorizontal wuiStepHorizontal-isIncomplete"
  role="tab"
  tabindex="0"
  title="Step 1: undefined"
>
  <div
    class="wuiScreenReaderOnly"
  >
    Step
  </div>
  <div
    class="wuiStepNumber wuiStepNumber--incomplete wuiStepHorizontal__number"
  >
    1
  </div>
  <div
    class="wuiStepHorizontal__title"
  />
</div>
`;

exports[`WuiStepHorizontal props status warning is rendered 1`] = `
<div
  aria-disabled="false"
  aria-selected="false"
  class="wuiStepHorizontal wuiStepHorizontal-isIncomplete"
  role="tab"
  tabindex="0"
  title="Step 1: undefined"
>
  <div
    class="wuiScreenReaderOnly"
  >
    Step
  </div>
  <div
    class="wuiStepNumber wuiStepNumber--warning wuiStepHorizontal__number"
  >
    <div
      aria-label="has warnings"
      class="wuiStepNumber__icon"
      data-wuiicon-type="alert"
    />
  </div>
  <div
    class="wuiStepHorizontal__title"
  />
</div>
`;

exports[`WuiStepHorizontal props step 1`] = `
<div
  aria-disabled="false"
  aria-selected="false"
  class="wuiStepHorizontal wuiStepHorizontal-isIncomplete"
  role="tab"
  tabindex="0"
  title="Step 5: undefined"
>
  <div
    class="wuiScreenReaderOnly"
  >
    Step
  </div>
  <div
    class="wuiStepNumber wuiStepNumber--incomplete wuiStepHorizontal__number"
  >
    5
  </div>
  <div
    class="wuiStepHorizontal__title"
  />
</div>
`;

exports[`WuiStepHorizontal props title 1`] = `
<div
  aria-disabled="false"
  aria-selected="false"
  class="wuiStepHorizontal wuiStepHorizontal-isIncomplete"
  role="tab"
  tabindex="0"
  title="Step 1: First step"
>
  <div
    class="wuiScreenReaderOnly"
  >
    Step
  </div>
  <div
    class="wuiStepNumber wuiStepNumber--incomplete wuiStepHorizontal__number"
  >
    1
  </div>
  <div
    class="wuiStepHorizontal__title"
  >
    First step
  </div>
</div>
`;
